### ISSUE TITLE
tf-2.0.0rc0-gpu tf.dataset bug

### ISSUE THREAD

**Issue by LaCandela** [tensorflow/tensorflow#31896]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code: yes 
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04 LTS
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): docker image
- TensorFlow version (use command below): tensorflow/tensorflow:2.0.0rc0-gpu-py3
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 10
- GPU model and memory:

**Describe the current behavior**
I am using tf.keras for training with the dataset api and TFrecords. At the beginning of the training the dataset api tries to load and shuffle the whole training dataset so the memory fills up and the process dies.  I cannot train my model at all. I am not using any .shuffle() calls in my dataset pipeline.

**Describe the expected behavior**
The dataset should not load and shuffle the records unless the .shuffle() call is issued.

**Code to reproduce the issue**
```
import numpy as np
import tensorflow as tf
import os
import cv2
import tensorflow.keras.layers as K

def extract_fn(data_record):
    features = {
        'data': tf.io.FixedLenFeature([], tf.string)
    }
    sample = tf.io.parse_single_example(data_record, features)
    data = tf.image.decode_image(sample['data'])

    return data, 1.


class DataGenerator(tf.keras.utils.Sequence):

    def __init__(self, dataset_iterator, len):
        self.dataset_iterator = dataset_iterator
        self.len = len

    def __len__(self):
        # number of batches per epoch
        return self.len

    def __getitem__(self, index):
        # Generate one batch of data

        next_element = next(self.dataset_iterator)
        x = next_element[0]
        y = next_element[1]

        return x, y


with tf.io.TFRecordWriter("dummy_dataset.tfrecords") as writer:
    data = np.float32(np.random.random(size=(1000, 1000, 3)) * 255)
    data = cv2.imencode(".png", data)[1].tostring()
    example = tf.train.Example(features=tf.train.Features(
        feature={'data': tf.train.Feature(bytes_list=tf.train.BytesList(value=[data]))}))
    for i in range(10000):
        writer.write(example.SerializeToString())

dataset = tf.data.TFRecordDataset(["dummy_dataset.tfrecords"])
dataset = dataset.map(extract_fn)
n_batch = 3
dataset = dataset.batch(batch_size=n_batch, drop_remainder=True)

dataset = dataset.repeat(5)
dataset_iterator = iter(dataset)
next_element = next(dataset_iterator)

data_generator = DataGenerator(dataset_iterator, int(10000/n_batch))

input = K.Input(shape=(1000, 1000, 3), name='input')
net = K.Conv2D(1, 3, activation='sigmoid')(input)
output = K.GlobalAveragePooling2D()(net)
model = tf.keras.models.Model(inputs=input, outputs=output)

model.compile(loss='mse', optimizer='sgd')

model.fit(x=data_generator, epochs=10)
```
**Other info / logs**
note that code generates about 28GB dummy data in a tfrecord file since I am having this issue when I am trying to use the dataset api with tf.keras.utils.Sequence on a tfrecord file.

EDIT(robieta): code formatting.

**Comment by oanush** [comment]
@LaCandela ,
In order to expedite the trouble-shooting process, please provide complete code snippet to reproduce the issue reported here. Thanks!

**Comment by LaCandela** [comment]
I found that the issue arises when I am trying to use tf.keras.utils.Sequence to create a data generator and using that to feed the model.fit. I provided a small script. Please note that it generates about 28GB dummy data in a tfrecord file since I am having this issue when I am trying to use the dataset api with tf.keras.utils.Sequence on a tfrecord file. I need to use the tf.keras.utils.Sequence because it allows me to do some preprocessing that I can performe on GPU. As best I know, I cannot do that with dataset.map().

**Comment by LaCandela** [comment]
Hi! I was wondering if there was any progress in this issue...

**Comment by jvishnuvardhan** [comment]
@LaCandela I reduced the dummy_data size and ran it with `TF2.0`. I don't see any issues. [Here](https://colab.sandbox.google.com/gist/jvishnuvardhan/007ed87a42cce8f0993933fd6c1c8b83/tf31896.ipynb) is the gist. Thanks!

**Comment by LaCandela** [comment]
@jvishnuvardhan Well, My TFrecords file size for the training is in the range of ~30GB and I cannot reduce that. The dataset api tries to fill a shuffle buffer which I am not calling at all (no dataset.shuffle() call) and quickly fills up my memory. It seems as if the whole TFrecords is loaded into memory. The issue is still present in TF2.0.

**Comment by jsimsa** [comment]
Why are using `DataGenerator` instead of passing your `dataset` directly to `model.fit`?

**Comment by LaCandela** [comment]
Because I want to do the preprocessing on the GPU and if I use the dataset.map() function the operations are performed on the CPU (as best I know). Now, I have the feeling that it is against the Tensorflow "mindset" but it would be great to have the possibility to experiment with the preproc (augmentation, ground truth generation) on the GPU approach.

**Comment by jsimsa** [comment]
The `map` dataset op will be placed on CPU but ops inside of the user-defined function passed into the `map` transformation can be placed on GPU.

As for the problem you are seeing, my best guess is that something in Keras internals might end up prefetching the contents of the entire generator ahead of time. @robieta could you please comment on what could be going on? Thanks.

**Comment by robieta** [comment]
I believe the issue is that shuffle defaults to True in model.fit, which will call .shuffle(len(sequence)) internally. (Prefetching all of the elements.) It does, however, strike me that for a Sequence this is not very efficient. (Since we can just directly index said sequence) I am already working on the data adapter; I will update it accordingly.

**Comment by saikumarchalla** [comment]
@LaCandela Could you check with latest TF version and let us know if the problem still persists or not.    Thanks!

**Comment by saikumarchalla** [comment]
@LaCandela  Any updates regarding this issue? Thanks!

**Comment by saikumarchalla** [comment]
@LaCandela  Closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
